### ISSUE TITLE
Layout/HashAlignment: allow grouping (i.e. separation by blank lines)

### ISSUE THREAD

**Issue by dmke** [rubocop/rubocop#11375]
## Is your feature request related to a problem? Please describe.

In a Rails project with JavaScript and Go code, I'd like to create a similar "shape" to the code, where possible.

To format my hashes, I'm currently using

```yaml
Layout/HashAlignment:
  EnforcedColonStyle:      table
  EnforcedHashRocketStyle: table
```

Code most affected by this usually are Rails' nested attributes (which require pretty long key names), and test code (where I might to create deeply nested hashes for FactoryBot). My controller code usually looks a bit like this:

```ruby
def permitted_params
  params.require(:user).permit(
    :email,
    :password,
    :password_confirmation,
    {
      avatar_attributes:              %i[
        color
        initials
        shape
        kind
        _destroy
      ],

      # after the blank line, a comment usually follows
      profile_preferences_attributes: %i[id key value],
    },
  )
end
```

Here, I find the spacing between `avatar_attributes:` and `%[` a little distracting.

## Describe the solution you'd like

ESlint allows to configure the `key-spacing` rule with [`{ "align": "value" }`](https://eslint.org/docs/latest/rules/key-spacing#align), which yields a similar output as Rubocop. Go (`gofmt` actually) is opinionated and doesn't allow for configuration, but also yields a similar output with struct fields and maps (click on "Format" [here](https://go.dev/play/p/_FSCP-Cp49y)).

Both ESlint and `gofmt` do apply different indentation level an a per-group basis, where "a group" are a set of keys and value separated by a blank line (ESlint says "non-consecutive lines form a new group").

```go
hash := map[string]string{
  "foo":    "42",
  "foobar": "string",

  "a_long_name": "b",
  "c":           "asdf",
}
```

```javascript
const hash = {
  foo:    "42",
  foobar: "string",

  a_long_name: "b",
  c:           "asdf",
}
```

I'd like to have a config knob for `Layout/HashAlignment`, which essentially does the same.

## Describe alternatives you've considered

The only alternatives currently are:

- live with the extra spacing, or
- disable the Layout cop and format hashes by hand.

## Additional context

In Go, there's actually a threshold for outliers, where this code doesn't get reformatted:

```go
hash := map[string]string{
  "foo":    "42",
  "foobar": "string",
  "a_very_unreasonably_long_name_with_many_letters": "b",
  "c":     "asdf",
  "yahoo": "news",
}
```

I'm not sure (a) what the threshold is, and (b) whether Rubocop should adopt this behaviour at all.

**Comment by dmke** [comment]
Still desired.

**Comment by dmke** [comment]
Desired, still.

**Comment by codezomb** [comment]
Here's a patch to `Layout/HashAlignment` that does this. I just did this as a quick solution in my own project. If there's interest I can clean it up more and put up a PR.

```ruby
# Patches Layout/HashAlignment to change EnforcedColonStyle: table behavior.
#
# Instead of aligning all colon-pair values to the widest key in the entire
# hash, each blank-line-separated group is aligned independently:
#
#   # good
#   {
#     foo:    'hi',
#     foobar: 'world',
#
#     a_long_key: 'b',
#     c:          'asdf',
#   }
#
# Hash-rocket pairs are unaffected by this patch.

# Patches Layout/HashAlignment EnforcedColonStyle: table to align colon-pair values
# per blank-line-separated group rather than across the full hash.
# Hash-rocket pairs fall through to the base cop unaffected.
#
#   # good — each group aligns independently
#   {
#     foo:    'hi',
#     foobar: 'world',
#
#     a_long_key: 'b',
#     c:          'asdf',
#   }

require 'rubocop'

module GroupedHashAlignment
  MSG = 'Align hash values within their blank-line-separated group.'.freeze

  def on_hash(node)
    return super unless cop_config['EnforcedColonStyle'] == 'table'
    return if node.single_line?

    return super if (pairs = grouped_colon_pairs(node)).nil?
    pairs.slice_when { |a, b| b.loc.line - a.loc.line > 1 }.each do |group|
      next if group.size < 2

      group.each do |pair|
        check_pair(pair, group.map { |p| key_op_width(p) }.max)
      end
    end
  end

  private

  def grouped_colon_pairs(node)
    return if node.pairs.size < 2 || node.pairs.each_cons(2).none? { |a, b| b.loc.line - a.loc.line > 1 }
    node.pairs.select { |p| p.colon? && p.key.loc.line == p.value.loc.line }
  end

  def check_pair(pair, width)
    return if pair.value.loc.column == pair.key.loc.column + width + 1

    add_offense(pair.value, message: MSG) do |corrector|
      corrector.replace(
        range_between(pair.loc.operator.end_pos, pair.value.loc.expression.begin_pos),
        ' ' * (width - key_op_width(pair) + 1)
      )
    end
  end

  def key_op_width(pair)
    pair.loc.operator.end_pos - pair.key.loc.expression.begin_pos
  end
end

RuboCop::Cop::Layout::HashAlignment
  .prepend(GroupedHashAlignment)
```
